File: frontend/src/app/core/services/credit.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { CreditSummary } from '../models/credit-summary.model';
import { Credit } from '../models/credit.model';
import { environment } from '../../../environments/environment';
import { StatutCredit } from '../models/enums/statut-credit.enum';

@Injectable({
  providedIn: 'root'
})
export class CreditService {
  private apiUrl = `${environment.apiUrl}/api/credits`;

  constructor(private http: HttpClient) { }

  getAllCredits(): Observable<CreditSummary[]> {
    return this.http.get<CreditSummary[]>(this.apiUrl);
  }

  getCreditById(id: number): Observable<Credit> {
    return this.http.get<Credit>(`${this.apiUrl}/${id}`);
  }

  createCredit(creditData: any): Observable<Credit> { // Use a more specific type e.g. CreditRequest
    return this.http.post<Credit>(this.apiUrl, creditData);
  }

  updateCredit(id: number, creditData: any): Observable<Credit> {
    return this.http.put<Credit>(`${this.apiUrl}/${id}`, creditData);
  }

  updateCreditStatus(id: number, statut: string): Observable<Credit> {
    return this.http.patch<Credit>(`${this.apiUrl}/${id}/statut`, { statut });
  }

  deleteCredit(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getCreditsByStatus(status: StatutCredit): Observable<CreditSummary[]> {
    return this.http.get<CreditSummary[]>(`${this.apiUrl}/status/${status}`);
  }

  getCreditsByType(type: string): Observable<CreditSummary[]> {
    return this.http.get<CreditSummary[]>(`${this.apiUrl}/type/${type}`);
  }

  approveCredit(id: number, approvalDate: Date): Observable<Credit> {
    return this.http.post<Credit>(`${this.apiUrl}/${id}/approve`, { approvalDate });
  }

  rejectCredit(id: number, reason: string): Observable<Credit> {
    return this.http.post<Credit>(`${this.apiUrl}/${id}/reject`, { reason });
  }

  calculateMonthlyPayment(id: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}/payment-calculation`);
  }

  getPaymentSchedule(id: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${id}/payment-schedule`);
  }

  validateCreditApplication(creditData: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/validate`, creditData);
  }

  searchCreditsByAmountRange(minAmount: number, maxAmount: number): Observable<CreditSummary[]> {
    return this.http.get<CreditSummary[]>(`${this.apiUrl}/search/amount`, {
      params: {
        minAmount: minAmount.toString(),
        maxAmount: maxAmount.toString()
      }
    });
  }

  searchCreditsByDateRange(startDate: Date, endDate: Date): Observable<CreditSummary[]> {
    return this.http.get<CreditSummary[]>(`${this.apiUrl}/search/date`, {
      params: {
        startDate: startDate.toISOString().split('T')[0],
        endDate: endDate.toISOString().split('T')[0]
      }
    });
  }
  countCredits(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/count`)
      .pipe(
        catchError(error => {
          console.error('Error counting credits:', error);
          return throwError(() => new Error(`Failed to count credits: ${error.message}`));
        })
      );
  }

  getCreditStatusSummary(): Observable<Map<StatutCredit, number>> {
    return this.http.get<Map<StatutCredit, number>>(`${this.apiUrl}/status-summary`)
      .pipe(
        catchError(error => {
          console.error('Error getting credit status summary:', error);
          return throwError(() => new Error(`Failed to get credit status summary: ${error.message}`));
        })
      );
  }
}

File: backend/src/main/java/com/souhailbektachi/backend/services/CreditService.java
package com.souhailbektachi.backend.services;

import com.souhailbektachi.backend.dtos.*;
import com.souhailbektachi.backend.entities.StatutCredit;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

/**
 * Service interface for operations related to credits
 */
public interface CreditService {

    /**
     * Retrieve all credits
     * 
     * @return List of all credits with summary information
     */
    List<CreditSummaryDTO> getAllCredits();

    /**
     * Get detailed information for a specific credit
     * 
     * @param id The credit's ID
     * @return Complete credit data including customer and repayment information
     * @throws RuntimeException if credit is not found
     */
    CreditDTO getCreditById(Long id);

    /**
     * Create a new credit of any type
     * 
     * @param creditRequestDTO The credit data to create (specific subtype)
     * @return The created credit with ID assigned
     * @throws RuntimeException if validation fails
     */
    CreditDTO createCredit(CreditRequestDTO creditRequestDTO);

    /**
     * Update an existing credit
     * 
     * @param id The credit's ID
     * @param creditRequestDTO The updated credit data
     * @return The updated credit information
     * @throws RuntimeException if credit is not found or validation fails
     */
    CreditDTO updateCredit(Long id, CreditRequestDTO creditRequestDTO);

    /**
     * Delete a credit by ID
     * 
     * @param id The credit's ID
     * @throws RuntimeException if credit is not found or cannot be deleted
     */
    void deleteCredit(Long id);

    /**
     * Filter credits by status
     * 
     * @param status The status to filter by
     * @return List of credits with the specified status
     */
    List<CreditSummaryDTO> getCreditsByStatus(StatutCredit status);
    
    /**
     * Filter credits by type
     * 
     * @param type The credit type: "PERSONNEL", "IMMOBILIER", "PROFESSIONNEL"
     * @return List of credits of the specified type
     */
    List<CreditSummaryDTO> getCreditsByType(String type);
    
    /**
     * Approve a credit application
     * 
     * @param id The credit's ID
     * @param approvalDate The date of approval
     * @return The updated credit with approved status
     * @throws RuntimeException if credit is not found or not in pending status
     */
    CreditDTO approveCredit(Long id, LocalDate approvalDate);
    
    /**
     * Reject a credit application
     * 
     * @param id The credit's ID
     * @param reason Reason for rejection (optional)
     * @return The updated credit with rejected status
     * @throws RuntimeException if credit is not found or not in pending status
     */
    CreditDTO rejectCredit(Long id, String reason);
    
    /**
     * Calculate monthly payment for a credit
     * 
     * @param id The credit's ID
     * @return Calculation result with payment details
     * @throws RuntimeException if credit is not found
     */
    Map<String, Object> calculateMonthlyPayment(Long id);
    
    /**
     * Get detailed payment schedule for a credit
     * 
     * @param id The credit's ID
     * @return List of scheduled payments with amounts and dates
     * @throws RuntimeException if credit is not found
     */
    List<Map<String, Object>> getPaymentSchedule(Long id);
    
    /**
     * Validate a new credit application
     * 
     * @param creditRequestDTO The credit data to validate
     * @return Validation result with success flag and any error messages
     */
    Map<String, Object> validateCreditApplication(CreditRequestDTO creditRequestDTO);
    
    /**
     * Search for credits by amount range
     * 
     * @param minAmount Minimum amount (optional)
     * @param maxAmount Maximum amount (optional)
     * @return List of credits within the specified amount range
     */
    List<CreditSummaryDTO> searchCreditsByAmountRange(Double minAmount, Double maxAmount);
    
    /**
     * Search for credits by date range
     * 
     * @param startDate Start date (optional)
     * @param endDate End date (optional)
     * @return List of credits within the specified date range
     */
    List<CreditSummaryDTO> searchCreditsByDateRange(LocalDate startDate, LocalDate endDate);
}
